// @sevenryze/ruid/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ruid = void 0;
const string_decoder_1 = require("string_decoder");
const validator_1 = __importDefault(require("validator"));
const uuid_1 = require("uuid");
class Ruid {
    constructor(ruidStringOrRuidBuffer) {
        if (Buffer.isBuffer(ruidStringOrRuidBuffer)) {
            const ruidString = Ruid.decodeIdBuffer(ruidStringOrRuidBuffer);
            if (!Ruid.checkRuidString(ruidString)) {
                throw new Error(`Invalid buffer type reference RUID.`);
            }
            this.ruidBuffer = ruidStringOrRuidBuffer;
            this.ruidString = ruidString;
        }
        else {
            if (ruidStringOrRuidBuffer && !Ruid.checkRuidString(ruidStringOrRuidBuffer)) {
                throw new Error(`Invliad string type reference RUID.`);
            }
            const ruidString = ruidStringOrRuidBuffer ? ruidStringOrRuidBuffer : Ruid.convertUuidToRuid(uuid_1.v1());
            this.ruidString = ruidString.toUpperCase();
            this.ruidBuffer = Buffer.from(ruidString, "hex");
        }
    }
    static getUuidV1() {
        return uuid_1.v1().toUpperCase();
    }
    static getUuidV4() {
        return uuid_1.v4().toUpperCase();
    }
    static transformToBuffer(id) {
        return Ruid.encodeIdString(id);
    }
    static transformToString(id) {
        return Ruid.decodeIdBuffer(id);
    }
    static checkRuidString(ruid) {
        return typeof ruid === "string" && ruid.length === 32 && validator_1.default.isHexadecimal(ruid);
    }
    isAfter(ruid) {
        const passingRuidBuffer = ruid.toBuffered();
        const result = this.ruidBuffer.compare(passingRuidBuffer);
        return result === 1;
    }
    isBefore(ruid) {
        const passingRuidBuffer = ruid.toBuffered();
        const result = this.ruidBuffer.compare(passingRuidBuffer);
        return result === -1;
    }
    isEqual(ruid) {
        const passingRuidBuffer = ruid.toBuffered();
        const result = this.ruidBuffer.compare(passingRuidBuffer);
        return result === 0;
    }
    fromUuid(uuid) {
        const ruidString = Ruid.convertUuidToRuid(uuid);
        return new Ruid(ruidString);
    }
    toString() {
        return this.ruidString;
    }
    toFriendlyString() {
        return `${this.ruidString.slice(0, 4)}-${this.ruidString.slice(4, 8)}-${this.ruidString.slice(8, 16)}-${this.ruidString.slice(16, 20)}-${this.ruidString.slice(20, 32)}`;
    }
    toJSON() {
        return this.ruidString;
    }
    toSqlString() {
        return `X'${this.ruidString}'`;
    }
    toBuffered() {
        return this.ruidBuffer;
    }
    static decodeIdBuffer(id) {
        const decoder = new string_decoder_1.StringDecoder("hex");
        return decoder.end(id).toUpperCase();
    }
    static encodeIdString(id) {
        return Buffer.from(id, "hex");
    }
    static convertUuidToRuid(uuid) {
        if (!validator_1.default.isUUID(uuid)) {
            throw new Error(`Invliad reference UUID.`);
        }
        const ruid = uuid.slice(14, 18) + uuid.slice(9, 13) + uuid.slice(0, 8) + uuid.slice(19, 23) + uuid.slice(24);
        return ruid.toUpperCase();
    }
}
exports.Ruid = Ruid;
