// @mastra/deployer/dist/chunk-EYS7MMO2.cjs
'use strict';

var chunkG4TSJE4W_cjs = require('./chunk-G4TSJE4W.cjs');
var chunkEXQIONBD_cjs = require('./chunk-EXQIONBD.cjs');
var chunkLG5W6D2R_cjs = require('./chunk-LG5W6D2R.cjs');
var rollup = require('rollup');
var path = require('path');
var resolveFrom = require('resolve-from');
var module$1 = require('module');
var logger = require('@mastra/core/logger');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var resolveFrom__default = /*#__PURE__*/_interopDefault(resolveFrom);

function isBuiltinModule(specifier) {
  return module$1.builtinModules.includes(specifier) || specifier.startsWith("node:") || module$1.builtinModules.includes(specifier.replace(/^node:/, ""));
}
function safeResolve(id, importer) {
  try {
    return resolveFrom__default.default(importer, id);
  } catch {
    return null;
  }
}
function getPackageName(id) {
  const parts = id.split("/");
  if (id.startsWith("@")) {
    return parts.slice(0, 2).join("/");
  }
  return parts[0];
}
function nodeModulesExtensionResolver() {
  return {
    name: "node-modules-extension-resolver",
    resolveId(id, importer) {
      if (id.startsWith(".") || id.startsWith("/") || !importer) {
        return null;
      }
      if (isBuiltinModule(id)) {
        return null;
      }
      if (id.startsWith("@") && id.split("/").length === 2) {
        return null;
      }
      if (!id.startsWith("@") && id.split("/").length === 1) {
        return null;
      }
      const foundExt = path.extname(id);
      if (foundExt) {
        return null;
      }
      try {
        const resolved = undefined(id);
        if (!path.extname(resolved)) {
          throw new Error(`Cannot resolve ${id} from ${importer}`);
        }
        return null;
      } catch (e) {
        const resolved = safeResolve(id, importer);
        if (resolved) {
          return {
            id: resolved,
            external: true
          };
        }
        for (const ext of [".mjs", ".js", ".cjs"]) {
          const resolved2 = safeResolve(id + ext, importer);
          if (resolved2) {
            const pkgName = getPackageName(id);
            if (!pkgName) {
              return null;
            }
            const pkgJsonPath = safeResolve(`${pkgName}/package.json`, importer);
            if (!pkgJsonPath) {
              return null;
            }
            const newImportWithExtension = resolved2.replace(path.dirname(pkgJsonPath), pkgName);
            return {
              id: newImportWithExtension,
              external: true
            };
          }
        }
      }
      return null;
    }
  };
}
async function getInputOptions2(entryFile, platform, env, { sourcemap = false, transpilePackages = [] } = {}) {
  const dependencies = /* @__PURE__ */ new Map();
  if (transpilePackages.length) {
    const { output, reverseVirtualReferenceMap } = await chunkG4TSJE4W_cjs.bundleExternals(
      new Map(transpilePackages.map((pkg) => [pkg, ["*"]])),
      ".mastra/.build",
      logger.noopLogger,
      {
        transpilePackages,
        isDev: true
      }
    );
    for (const file of output) {
      if (file.type === "asset") {
        continue;
      }
      if (file.isEntry && reverseVirtualReferenceMap.has(file.name)) {
        dependencies.set(reverseVirtualReferenceMap.get(file.name), file.fileName);
      }
    }
  }
  const inputOptions = await chunkEXQIONBD_cjs.getInputOptions(
    entryFile,
    {
      dependencies,
      externalDependencies: /* @__PURE__ */ new Set(),
      invalidChunks: /* @__PURE__ */ new Set()
    },
    platform,
    env,
    { sourcemap }
  );
  if (Array.isArray(inputOptions.plugins)) {
    const plugins = [];
    inputOptions.plugins.forEach((plugin) => {
      if (plugin?.name === "node-resolve") {
        return;
      }
      if (plugin?.name === "tsconfig-paths") {
        plugins.push(
          chunkLG5W6D2R_cjs.tsConfigPaths({
            localResolve: true
          })
        );
        return;
      }
      plugins.push(plugin);
    });
    inputOptions.plugins = plugins;
    inputOptions.plugins.push(chunkG4TSJE4W_cjs.aliasHono());
    inputOptions.plugins.push(nodeModulesExtensionResolver());
  }
  return inputOptions;
}
async function createWatcher(inputOptions, outputOptions) {
  const watcher = await rollup.watch({
    ...inputOptions,
    output: {
      ...outputOptions,
      format: "esm",
      entryFileNames: "[name].mjs",
      chunkFileNames: "[name].mjs"
    }
  });
  return watcher;
}

// src/build/babel/remove-all-options-server.ts
function removeAllOptionsExceptServer(result, logger) {
  return chunkG4TSJE4W_cjs.removeAllOptionsFromMastraExcept(result, "server", logger);
}

// src/build/serverOptions.ts
async function getServerOptions(entryFile, outputDir, logger) {
  const result = await chunkG4TSJE4W_cjs.extractMastraOption(
    "server",
    entryFile,
    removeAllOptionsExceptServer,
    outputDir,
    logger
  );
  if (!result) {
    return null;
  }
  return result.getConfig();
}

exports.createWatcher = createWatcher;
exports.getInputOptions = getInputOptions2;
exports.getServerOptions = getServerOptions;
