// parse-jsx-to-css/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parentHasSamePrefixChild = exports.classNameExtendParent = exports.getOutputFullPath = exports.collectionVueClassNames = exports.collectAllReactClassNames = void 0;
function isExpressionNode(pet) {
    if (!pet) {
        return false;
    }
    return pet.type === "JSXExpressionContainer";
}
function getExpressionNodeAttrs(node) {
    if (node.expression.type === "ConditionalExpression") {
        // 针对表达式的情况，则返回表达式两个
        var result = [];
        var conditionNode = node.expression;
        if (conditionNode.consequent.type === "JSXElement") {
            result = result.concat(conditionNode.consequent.openingElement.attributes);
        }
        if (conditionNode.alternate.type === "JSXElement") {
            result = result.concat(conditionNode.alternate.openingElement.attributes);
        }
        return result;
    }
    return [];
}
function collectAllReactClassNames(node) {
    var _a, _b, _c;
    var itemJSX = {
        name: "",
        children: [],
    };
    if (isExpressionNode(node)) {
        if (node.expression.type === "ConditionalExpression") {
            var conditionNode = node.expression;
            if (conditionNode.consequent.type === "JSXElement") {
                (_a = itemJSX.children) === null || _a === void 0 ? void 0 : _a.push(collectAllReactClassNames(conditionNode.consequent));
            }
            if (conditionNode.alternate.type === "JSXElement") {
                (_b = itemJSX.children) === null || _b === void 0 ? void 0 : _b.push(collectAllReactClassNames(conditionNode.alternate));
            }
            return itemJSX;
        }
    }
    else {
        var attrs = node.openingElement.attributes;
        for (var index = 0; index < attrs.length; index++) {
            var ele = attrs[index];
            if (ele.type === "JSXAttribute") {
                if (ele.name.name === "className" &&
                    ((_c = ele.value) === null || _c === void 0 ? void 0 : _c.type) === "StringLiteral") {
                    itemJSX.name = ele.value.value;
                }
            }
        }
        node.children.forEach(function (ele) {
            var _a;
            if (ele.type === "JSXElement" || ele.type === "JSXExpressionContainer") {
                (_a = itemJSX.children) === null || _a === void 0 ? void 0 : _a.push(collectAllReactClassNames(ele));
            }
        });
    }
    return itemJSX;
}
exports.collectAllReactClassNames = collectAllReactClassNames;
/**
 * is vue attribute node?
 * @param pet
 */
function isAttributeNode(pet) {
    if (!pet) {
        return false;
    }
    return pet.value !== undefined;
}
/**
 * is vue TemplateChildNode node?
 * @param pet
 */
function isTemplateChildNode(node) {
    if (!node) {
        return false;
    }
    if (node && node.type === 1 /* ELEMENT */ && node.tag === "template") {
        return true;
    }
    return false;
}
/**
 * return the vue nodes
 * @param vueNode
 */
function collectionVueClassNames(vueNode) {
    var _a;
    var itemJSX = {
        name: "",
        children: [],
    };
    if (vueNode.type === 1 /* ELEMENT */ || vueNode.type === 0 /* ROOT */) {
        if (vueNode.type === 1 /* ELEMENT */ &&
            vueNode.tag !== "script" &&
            vueNode.tag !== "style") {
            var classNameNode = vueNode.props.find(function (ele) { return ele.name === "class"; });
            if (isAttributeNode(classNameNode)) {
                itemJSX.name = (_a = classNameNode.value) === null || _a === void 0 ? void 0 : _a.content;
            }
        }
        vueNode.children.forEach(function (templateNode) {
            var _a;
            if (templateNode.type === 1 /* ELEMENT */ &&
                templateNode.tag !== "script" &&
                templateNode.tag !== "style") {
                var next = collectionVueClassNames(templateNode);
                (_a = itemJSX.children) === null || _a === void 0 ? void 0 : _a.push(next);
            }
        });
    }
    return itemJSX;
}
exports.collectionVueClassNames = collectionVueClassNames;
/**
 * return the file save path
 * @param options
 */
function getOutputFullPath(options) {
    var path = options.input.substring(0, options.input.lastIndexOf(".") + 1);
    return path + (options.outType === "sass" ? "scss" : options.outType);
}
exports.getOutputFullPath = getOutputFullPath;
/**
 * child class name is extends parent
 * @param className
 * @param parent
 */
function classNameExtendParent(className, parent) {
    for (var index = 0; index < parent.length; index++) {
        var ele = parent[index];
        if (className.startsWith(ele)) {
            return [ele, index];
        }
    }
    return ["", -1];
}
exports.classNameExtendParent = classNameExtendParent;
/**
 * parent has same prefix child
 * @param parentName
 * @param children
 */
function parentHasSamePrefixChild(parentName, itemClass) {
    if (itemClass.children && Array.isArray(itemClass.children)) {
        for (var i = 0; i < itemClass.children.length; i++) {
            var element = itemClass.children[i].name || "";
            var _a = classNameExtendParent(element, parentName), parent_1 = _a[0], index = _a[1];
            if (parent_1 && index !== -1) {
                return [true, index];
            }
        }
    }
    return [false, -1];
}
exports.parentHasSamePrefixChild = parentHasSamePrefixChild;
