// @mastra/server/dist/server/handlers/legacyWorkflows.cjs
'use strict';

var chunkX7NUY6QT_cjs = require('../../chunk-X7NUY6QT.cjs');



Object.defineProperty(exports, "createLegacyWorkflowRunHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.createLegacyWorkflowRunHandler; }
});
Object.defineProperty(exports, "getLegacyWorkflowByIdHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.getLegacyWorkflowByIdHandler; }
});
Object.defineProperty(exports, "getLegacyWorkflowRunHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.getLegacyWorkflowRunHandler; }
});
Object.defineProperty(exports, "getLegacyWorkflowRunsHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.getLegacyWorkflowRunsHandler; }
});
Object.defineProperty(exports, "getLegacyWorkflowsHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.getLegacyWorkflowsHandler; }
});
Object.defineProperty(exports, "resumeAsyncLegacyWorkflowHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.resumeAsyncLegacyWorkflowHandler; }
});
Object.defineProperty(exports, "resumeLegacyWorkflowHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.resumeLegacyWorkflowHandler; }
});
Object.defineProperty(exports, "startAsyncLegacyWorkflowHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.startAsyncLegacyWorkflowHandler; }
});
Object.defineProperty(exports, "startLegacyWorkflowRunHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.startLegacyWorkflowRunHandler; }
});
Object.defineProperty(exports, "watchLegacyWorkflowHandler", {
  enumerable: true,
  get: function () { return chunkX7NUY6QT_cjs.watchLegacyWorkflowHandler; }
});
